### ISSUE TITLE
Inconsistent behaviour of `Unpack` with `TypedDict(name, dict)`

### ISSUE THREAD

**Issue by jakelevi1996** [microsoft/pyright#9125]
## Environment data

-   Pylance version: Pylance  v2024.9.2 
-   OS and version: Ubuntu 22.04.1 LTS
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.10.12

## Code Snippet

```python
from typing_extensions import TypedDict, Unpack

kwarg_type_dict = {'k1': str, 'k2': str}
t1 = TypedDict("FKwargs1", kwarg_type_dict)
def f1(**kwargs: Unpack[t1]): pass

t2 = TypedDict("FKwargs2", {'k1': str, 'k2': str})
def f2(**kwargs: Unpack[t2]): pass
```

## Repro Steps

1. Save the above code in a Python file
2. Type `f1()`, observe type hint
3. Type `f2()`, observe type hint

## Expected behavior

`f1` and `f2` have the same correct type hints.

## Actual behavior

`f2` has the correct type hint, `f1` is wrong:

![image](https://github.com/user-attachments/assets/ce0330ab-bd69-4899-b449-702f90617350)

![image](https://github.com/user-attachments/assets/fd9f4ffe-46c7-4077-855a-e06526f802c6)

## Aside

FYI, the reason I care about this is a use case in which I forward `**kwargs` to another function, and can extract those target `kwarg` names and types using the `inspect` module - but I can't add these as a type hint that shows up in VS Code, for the reason mentioned above. Example shown below. It would be nice if this use case is supported, but just to be clear, the bug I am reporting is the one I mentioned above.

```python
import inspect
from typing_extensions import TypedDict, Unpack

def f1(a1: int, k1: str="v1", k2: str="v2") -> None:
    print("f1", a1, k1, k2)

def f2(a2: int) -> int:
    print("f2", a2)
    return a2 + 1

kwarg_type_dict: dict[str, type] = {
    p.name: p.annotation
    for p in inspect.signature(f1).parameters.values()
    if p.default is not inspect.Parameter.empty
}
t = TypedDict("F3Kwargs", kwarg_type_dict)

def f3(a3: int, **kwargs: Unpack[t]) -> None:
    f1((f2(a3)), **kwargs)
```

## Logs

```
XXX
```

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug.

The behavior of `TypedDict` is specified in the Python typing spec, and pyright is conformant with that spec. See [this section](https://typing.readthedocs.io/en/latest/spec/typeddict.html#alternative-syntax) for details.

If you enable type checking errors in pylance (i.e. set `typeCheckingMode` to "basic"), you will see that your code is using the `TypedDict` constructor incorrectly here.

**Comment by jakelevi1996** [comment]
Thanks @erictraut , got it. Didn't realise the dictionary couldn't be a variable.

On this topic, in the "aside" I mentioned above, if I know exactly the names and types (and defaults) of all the options for `**kwargs`, and can extract them into a dictionary using `inspect.signature`, is there anyway I can specify this to Python/Pylance/Pyright , so that I can get the right type hints for them in VS Code?

(of course I could just copy+paste, but I want them to stay in sync after making changes without having to copy and paste every time)

(I also tried modifying `__signature__` for the function directly, shown in [this Gist](https://gist.github.com/jakelevi1996/f63006521d17153c6e49d5ec5e6f4c97), and while this modified the behaviour correctly for `inspect.signature`, it didn't work for getting the correct type hints in VS Code)

**Comment by erictraut** [comment]
Pyright is a static type analyzer. It doesn't run any of your Python code. Extracting dictionary information through dynamic calls to `inspect.signature` will work at runtime, but it won't work with static analysis tools. For static analysis, you need to follow the rules in the Python typing spec.

**Comment by jakelevi1996** [comment]
A simpler solution (if it existed, I'm not sure if it does), rather than having to modify the build process to auto-generate an interface file and keep it up to date on VCS etc, is if I could specify a `ParamSpec` that was bound to a specific function, and then just extract the `.kwargs` attribute from that `ParamSpec` and use it in the type hint for `**kwargs` in the wrapping function - something like this (with `ParamSpec.from_function`):

```python
from typing_extensions import ParamSpec

def f1(k1: str="v1", k2: str="v2") -> None:
    print("f1", k1, k2)

def f2(a3: int, **kwargs: ParamSpec.from_function[f1].kwargs) -> None:
    print("f2", a3)
    f1(**kwargs)

f2(3, k2="v22")
```

Do you know if there's a way to do this, or plans to support this in `typing`? I've seen similar issues approached using typed decorators, but they seem to require a lot of boilerplate relative to the simplicity of what they're trying to achieve (and less readable, etc). Would be convenient if there was a way to do this like above

**Comment by erictraut** [comment]
> Do you know if there's a way to do this

No, the [Python static type system](https://typing.readthedocs.io/en/latest/spec/index.html) doesn't support this. 

I don't think your proposed solution would work for a number of reasons. One reason is that in Python, most parameters can be specified using position or keyword arguments. In your example above, `k1` and `k2` are examples of this. The only way to determine whether `k1` and `k2` are kwargs is to look at a caller, and it can vary from one caller to another. This is why `P.args` and `P.kwargs` always need to appear together in a signature.

**Comment by jakelevi1996** [comment]
> This is why P.args and P.kwargs always need to appear together in a signature

Could it work if it was something like this?

```python

from typing_extensions import ParamSpec

def f1(k1: str="v1", k2: str="v2") -> None:
    print("f1", k1, k2)

P = ParamSpec("P", from_function=f1)

def f2(a3: int, *args: P.args, **kwargs: P.kwargs) -> None:
    print("f2", a3)
    f1(**kwargs)

f2(3, k2="v22")
```

**Comment by jakelevi1996** [comment]
I see... it does feel like a shame, I think there are definitely situations where it is useful to accept various positional and keyword arguments as well as variadic `**kwargs`, with `**kwargs` simply being forwarded to a specific function with known signature, and it's a shame not be able to get type hints/VS Code suggestions for `**kwargs` (this is done all the time in `matplotlib` source code, and I sometimes use it in [my own code](https://github.com/jakelevi1996/jutility/blob/main/src/jutility/plotting.py#L723))
